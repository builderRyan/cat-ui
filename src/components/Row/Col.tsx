import * as React from 'react';
import { PureComponent } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

export interface ResponsiveProps {
  span?: number;
  offset?: number;
  order?: number;
}

export interface ColProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 栅格所占的格数
   * 当值为 0 时，display: none; 隐藏
   */
  span?: number;
  /**
   * 左偏移的格数
   */
  offset?: number;
  /**
   * 栅格的顺序
   */
  order?: number;
  /**
   * width < 576px 时的其他属性对象
   */
  xs?: ResponsiveProps;
  /**
   * width >= 576px 时的其他属性对象
   */
  sm?: ResponsiveProps;
  /**
   * width >= 768px 时的其他属性对象
   */
  md?: ResponsiveProps;
  /**
   * width >= 992px 时的其他属性对象
   */
  lg?: ResponsiveProps;
  /**
   * width >= 1200px 时的其他属性对象
   */
  xl?: ResponsiveProps;
  /**
   * width >= 1200px 时的其他属性对象
   */
  xxl?: ResponsiveProps;
  /**
   * width >= 1920px 时的其他属性对象
   */
  fhd?: ResponsiveProps;
}

/**
 * Col
 */
class Col extends PureComponent<ColProps> {
  static propTypes = {
    className: PropTypes.string,
    span: PropTypes.number,
    offset: PropTypes.number,
    xs: PropTypes.object,
    sm: PropTypes.object,
    md: PropTypes.object,
    lg: PropTypes.object,
    xl: PropTypes.object,
    xxl: PropTypes.object,
    fhd: PropTypes.object
  };

  static defaultProps = {
    span: 0,
    offset: 0
  };

  getClasses = () => {
    const {
      className,
      span,
      offset,
      order,
      xs,
      sm,
      md,
      lg,
      xl,
      xxl,
      fhd
    } = this.props;

    const getMediaClasses = (
      sizeName: string,
      responsiveProps?: ResponsiveProps
    ) => {
      const prefix = 'cat-col';
      return {
        [`${prefix}--${sizeName}-span-${
          typeof responsiveProps?.span === 'number' ? responsiveProps?.span : 0
        }`]: responsiveProps,
        [`${prefix}--${sizeName}-offset-${
          typeof responsiveProps?.offset === 'number'
            ? responsiveProps?.offset
            : 0
        }`]: responsiveProps,
        [`${prefix}--${sizeName}-order-${
          typeof responsiveProps?.order === 'number'
            ? responsiveProps?.order
            : 0
        }`]: responsiveProps?.order
      };
    };

    const prefix = `cat-col`;
    const classes = classNames(
      prefix,
      `${prefix}--span-${span}`,
      `${prefix}--offset-${offset}`,
      {
        [`${prefix}--order-${order}`]: order,
        ...getMediaClasses('xs', xs),
        ...getMediaClasses('sm', sm),
        ...getMediaClasses('md', md),
        ...getMediaClasses('lg', lg),
        ...getMediaClasses('xl', xl),
        ...getMediaClasses('xxl', xxl),
        ...getMediaClasses('fhd', fhd)
      },
      className
    );
    return classes;
  };

  render() {
    const { children } = this.props;

    const classes = this.getClasses();

    return <div className={classes}>{children}</div>;
  }
}

export default Col;
