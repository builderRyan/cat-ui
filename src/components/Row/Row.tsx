import * as React from 'react';
import { Component } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import Col from './Col';

type AlignType = 'top' | 'center' | 'bottom';

type JustifyType =
  | 'start'
  | 'end'
  | 'center'
  | 'space-around'
  | 'space-between';

export interface RowProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * flex 布局下的水平排列方式
   */
  justify?: JustifyType;
  /**
   * flex 布局下的垂直对齐方式
   */
  align?: AlignType;
}

enum AlignEnum {
  top = 'start',
  center = 'center',
  bottom = 'bottom'
}

/**
 * Row
 */
class Row extends Component<RowProps> {
  static Col: typeof Col;
  static propTypes = {
    className: PropTypes.string,
    justify: PropTypes.oneOf([
      'start',
      'end',
      'center',
      'space-around',
      'space-between'
    ]),
    align: PropTypes.oneOf(['top', 'center', 'bottom'])
  };

  static defaultProps = {
    justify: 'start',
    align: 'top'
  };

  render() {
    const { children, className, justify, align } = this.props;

    const prefix = 'cat-row';
    const classes = classNames(
      prefix,
      className,
      `${prefix}--justify-${justify}`,
      `${prefix}--align-${AlignEnum[align]}`
    );

    return <div className={classes}>{children}</div>;
  }
}

export default Row;
