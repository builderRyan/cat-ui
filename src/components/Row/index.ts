import Row from './Row';
import Col from './Col';
Row.Col = Col;

export default Row;
