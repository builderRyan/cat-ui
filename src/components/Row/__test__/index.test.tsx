import * as React from 'react';
import { render } from 'enzyme';
import Row from '../index';

const { Col } = Row;

describe('Button', () => {
  test('renders correctly', () => {
    const wrapper1 = render(
      <Row>
        <Col span={12}>col-12</Col>
        <Col span={12}>col-12</Col>
      </Row>
    );
    expect(wrapper1).toMatchSnapshot();
    const wrapper2 = render(
      <Row>
        <Col xs={{ span: 6, order: 1 }} sm={{ span: 6, order: 2, offset: 1 }}>
          col-12
        </Col>
        <Col xs={{ span: 6, offset: 1, order: 2 }} sm={{ span: 6, order: 1 }}>
          col-12
        </Col>
      </Row>
    );
    expect(wrapper2).toMatchSnapshot();
  });
});
