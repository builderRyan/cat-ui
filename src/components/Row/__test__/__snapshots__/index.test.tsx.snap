// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div
  class="cat-row cat-row--justify-start cat-row--align-start"
>
  <div
    class="cat-col cat-col--span-12 cat-col--offset-0"
  >
    col-12
  </div>
  <div
    class="cat-col cat-col--span-12 cat-col--offset-0"
  >
    col-12
  </div>
</div>
`;

exports[`Button renders correctly 2`] = `
<div
  class="cat-row cat-row--justify-start cat-row--align-start"
>
  <div
    class="cat-col cat-col--span-0 cat-col--offset-0 cat-col--xs-span-6 cat-col--xs-offset-0 cat-col--xs-order-1 cat-col--sm-span-6 cat-col--sm-offset-1 cat-col--sm-order-2"
  >
    col-12
  </div>
  <div
    class="cat-col cat-col--span-0 cat-col--offset-0 cat-col--xs-span-6 cat-col--xs-offset-1 cat-col--xs-order-2 cat-col--sm-span-6 cat-col--sm-offset-0 cat-col--sm-order-1"
  >
    col-12
  </div>
</div>
`;
